src/frontend/src/binder/expr/function.rs

```diff
@@ -117,8 +117,14 @@ impl Binder {
             let digits = inputs.pop().unwrap();
             let input = inputs.pop().unwrap();
             vec![
-                input.ensure_type(DataType::Decimal),
-                digits.ensure_type(DataType::Int32),
+                input
+                    .clone()
+                    .cast_implicit(DataType::Decimal)
```

Comment:
How about making `cast_xxx` returns the `ExprImpl` itself when casting failed? So that we can avoid this `clone` here with `unwrap_or_else`.

For example, `Result<ExprImpl, (RwError, ExprImpl)>`.

Reply:
Good idea. But the implicit cast implementation here may not be final. The current implementation is a little bit hacky/adhoc and may be refactored to a more declarative way to support more functions.

Basically, the original expr is retained on error just to allow `infer_type` to report an error message with proper types. We already know it is going to fail for sure. This is the only place we are doing implicit cast on operands, as a pioneer. It does need a refactor to remove this hack, but `Result<ExprImpl, (RwError, ExprImpl)>` is not the final answer either.